Content.Server/MachineLinking/Models/SignalLinkCollection.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System.Collections.Generic;
+using Content.Server.MachineLinking.Components;
+using Content.Server.MachineLinking.Exceptions;
+
+namespace Content.Server.MachineLinking.Models
+{
+    public class SignalLinkCollection
+    {
+        private Dictionary<SignalTransmitterComponent, List<SignalLink>> _transmitterDict = new();
+        private Dictionary<SignalReceiverComponent, List<SignalLink>> _receiverDict = new();
+
+        public SignalLink AddLink(SignalTransmitterComponent transmitterComponent, string transmitterPort,
+            SignalReceiverComponent receiverComponent, string receiverPort)
+        {
+            if (LinkExists(transmitterComponent, transmitterPort, receiverComponent, receiverPort))
+            {
+                throw new LinkAlreadyRegisteredException();
+            }
+
+            if (!_transmitterDict.ContainsKey(transmitterComponent))
+            {
+                _transmitterDict[transmitterComponent] = new();
+            }
+
+            if (!_receiverDict.ContainsKey(receiverComponent))
+            {
+                _receiverDict[receiverComponent] = new();
+            }
+
+            var link = new SignalLink(transmitterComponent, transmitterPort, receiverComponent, receiverPort);
+            _transmitterDict[transmitterComponent].Add(link);
+            _receiverDict[receiverComponent].Add(link);
+
+            return link;
+        }
+
+        public bool LinkExists(SignalTransmitterComponent transmitterComponent, string transmitterPort,
+            SignalReceiverComponent receiverComponent, string receiverPort)
+        {
+            if (!_transmitterDict.ContainsKey(transmitterComponent) || !_receiverDict.ContainsKey(receiverComponent))
+            {
+                return false;
+            }
+
+            foreach (var link in _transmitterDict[transmitterComponent])
+            {
+                if (link.Transmitterport.Name == transmitterPort && link.Receiverport.Name == receiverPort &&
+                    link.ReceiverComponent == receiverComponent)
+                    return true;
+            }
+
+            return false;
+        }
+
+        public bool RemoveLink(SignalTransmitterComponent transmitterComponent, string transmitterPort,
+            SignalReceiverComponent receiverComponent, string receiverPort)
+        {
+            if (!_transmitterDict.ContainsKey(transmitterComponent) || !_receiverDict.ContainsKey(receiverComponent))
+            {
+                return false;
+            }
+
+            SignalLink? theLink = null;
+            foreach (var link in _transmitterDict[transmitterComponent])
+            {
+                if(link.Transmitterport.Name == transmitterPort && link.Receiverport.Name == receiverPort &&
```

Comment:
```suggestion
                if (link.Transmitterport.Name == transmitterPort && link.Receiverport.Name == receiverPort &&
```
Bunch of these in here.